Feature: Pegin proofs v1

SatwikPrabhu opened the pull request:
Pegin proof v1 based on the discussion in https://github.com/botanix-labs/botanix/issues/909

scottmillner committed:
refactor(btc-server): only untrack if error message confirms tx does not exist

SatwikPrabhu commented on crates/consensus/authority/src/excecution_utils.rs:
```diff
@@ -347,12 +349,25 @@ pub(crate) mod authority_execution_utils {
+        let blockchain_db = Arc::new(BlockchainProvider2::new(
+            ProviderFactory::new(
+                Arc::new(DatabaseEnv::open(
+                    std::path::Path::new("/tmp/reth_db"),
```
Was not sure what path to use. Would appreciate any suggestions.

SatwikPrabhu commented on crates/ethereum/evm/src/execute.rs:
```diff
@@ -64,17 +67,29 @@ pub struct EthExecutorProvider<BF, EvmConfig = EthEvmConfig> {
+        Arc::new(BlockchainProvider2::new(
+            ProviderFactory::new(
+                Arc::new(DatabaseEnv::open(
+                    std::path::Path::new("/tmp/reth_db"),
```
Same as https://github.com/botanix-labs/Macbeth/pull/610/files#r1982262094

SatwikPrabhu commented on crates/ethereum/node/src/node.rs:
```diff
@@ -122,8 +126,19 @@ where
-        let executor =
-            EthExecutorProvider::new(chain_spec, evm_config, mock_bitcoind_factory, regtest);
+        let db = Arc::new(DatabaseEnv::open(
+            std::path::Path::new("/tmp/reth_db"),
```
Same as https://github.com/botanix-labs/Macbeth/pull/610/files#r1982262094

arminsabouri commented on bin/reth/src/commands/poa/mod.rs:
```diff
@@ -527,14 +527,7 @@ impl<Ext: clap::Args + fmt::Debug> PoaNodeCommand<Ext> {
         let sync_metrics_listener = reth_stages::MetricsListener::new(sync_metrics_rx);
         executor.spawn_critical("stages metrics listener task", sync_metrics_listener);
 
-        // Config executor factory
-        let evm_config = EthEvmConfig::default();
-        let executor_factory = EthExecutorProvider::new(
-            Arc::new(chain.clone()),
-            evm_config,
-            bitcoind_factory.clone(),
-            node_config.rpc.btc_network,
-        );
```
Did this need to get moved? Please revert if not

SatwikPrabhu replied:
It does need to because we are initializing the db provider later

arminsabouri replied:
Ah i see. But if https://github.com/botanix-labs/Macbeth/pull/610#discussion_r1982428395 ends up working then we can revert this

arminsabouri commented on crates/consensus/authority/src/excecution_utils.rs:
```diff
@@ -347,12 +349,25 @@ pub(crate) mod authority_execution_utils {
             .with_bundle_update()
             .build();
 
+        let blockchain_db = Arc::new(BlockchainProvider2::new(
```
can we not use the provider that is defined above this?

SatwikPrabhu replied:
I tried doing this. The issue here is that we use `reth_db::Database` in the authority consensus crate and `revm_primitives::db::Database` in the ethereum executor code. The former is clonable but the latter is not among other differences

arminsabouri commented on crates/consensus/authority/src/excecution_utils.rs:
```diff
@@ -347,12 +349,25 @@ pub(crate) mod authority_execution_utils {
+                    std::path::Path::new("/tmp/reth_db"),
+                    reth_db::DatabaseEnvKind::RO,
+                    reth_db::mdbx::DatabaseArguments::default()
+                ).expect("Failed to open database")),
```
Should not panic here

SatwikPrabhu replied:
Done!

arminsabouri commented on crates/consensus/authority/src/excecution_utils.rs:
```diff
@@ -347,12 +349,25 @@ pub(crate) mod authority_execution_utils {
+                chain_spec.clone(),
+                StaticFileProvider::default()
+            )
+        ).expect("Failed to create blockchain provider"));
```
Same here

SatwikPrabhu replied:
done!

arminsabouri commented on crates/ethereum/evm/src/execute.rs:
```diff
@@ -64,17 +67,29 @@ pub struct EthExecutorProvider<BF, EvmConfig = EthEvmConfig> {
         EthEvmConfig::default(),
         MockBitcoindFactory::new(BitcoindConfig::default()),
         bitcoin::Network::Regtest,
+        Arc::new(BlockchainProvider2::new(
```
There are noop providers within the reth project

SatwikPrabhu replied:
I did not find any for `BlockchainProvider2`

arminsabouri commented on crates/ethereum/evm/src/execute.rs:
```diff
@@ -426,7 +451,38 @@ where
+            // The reference block hash is only provided for version v1.
+            if pegin_data.meta[0].version == 1 && pegin_data.meta[0].ref_block_hash.is_none() {
+                return Err(MintContractError::InvalidPeginData {
+                    error: format!("Reference block hash not found"),
```
```suggestion
                    error: format!("Reference block hash cannot be none"),
```

SatwikPrabhu replied:
Done!

arminsabouri commented on crates/ethereum/evm/src/execute.rs:
```diff
@@ -426,7 +451,38 @@ where
+            // This is used to avoid the growing list of headers in the pegin metadata
+            // by using a bitcoin checkpoint that is close to the pegin block height. 
+            // The reference block hash is only provided for version v1.
+            if pegin_data.meta[0].version == 1 && pegin_data.meta[0].ref_block_hash.is_none() {
```
why index into 0? Wouldnt be possible that meta could be erroneously empty? And I think we should be iterating through all the proofs, not just the first

SatwikPrabhu replied:
Let me know if its fine now

arminsabouri commented on crates/ethereum/evm/src/execute.rs:
```diff
@@ -426,7 +451,38 @@ where
+
+            let bitcoin_checkpoint = if let Some(ref_block_hash) = pegin_data.meta[0].ref_block_hash {
+                if let Some(block) = provider.find_block_by_hash(ref_block_hash, reth_provider::BlockSource::Any).map_err(|_| MintContractError::InvalidPeginData {
+                    error: format!("Block not found"),
```
Lets be more specific.
```suggestion
                    error: format!("refrence block hash not found"),
```

SatwikPrabhu replied:
Done!

arminsabouri commented on crates/primitives/src/botanix/peg_contract.rs:
```diff
@@ -150,6 +151,8 @@ pub struct PeginMeta {
     /// Pegin tx
     pub tx: bitcoin::Transaction,
+    /// Reference botanix block hash whose checkpoint is used for validating the pegin
+    pub ref_block_hash: Option<B256>,
```
In my opinion there is a more ergonomic design to consider here. A v1 pegin proof includes all the fields of a v0 proof. As such you can create a v1 meta struct that just encapsulates the v0 struct.

```rust 

pub struct PeginMetaV1 {
v0: ...,
ref_block_hash: Option<B256>

}
```

arminsabouri replied:
This may require some refactoring, as PeginMeta is really reperesenting what is serialized and deserialized and the suggestion above implies there is another layer of abstraction.

SatwikPrabhu replied:
I have given this a try. Let me know how it looks.

arminsabouri commented on crates/ethereum/evm/src/execute.rs:
```diff
@@ -223,6 +243,8 @@ where
         let base_fee = block.base_fee_per_gas;
         let mut receipts = Vec::with_capacity(block.body.len());
 
+        let provider = self.blockchain_db.clone();
```
The evm object passed to this function appears to have access to the databse 
https://github.com/botanix-labs/Macbeth/pull/610/files#diff-7fbb0c1e65c92e8879aa308fee2538af673a2cd6d70f9b5b1103cf4d6554604dL210

Is passing the db all the way down here neccecary?

SatwikPrabhu replied:
I tried doing this actually. But it seems the evm object is of type `State<DB>' and does not implement `BlockReader`. So, this cannot do block retrieval from the db.

arminsabouri replied:
Ok. Can we try passing in the provider when calling this method. Similar to how evm is passed in. That should reduce the codefoot print a bit and follows the current pattern of passing stateful things into the method not top down

SatwikPrabhu replied:
Done!

arminsabouri reviewed (changes requested):
This is on the right track and I commend you for pioneering us into a new pegin version! But I did notice a couple things that stuck out and since this is touching probably the most critical consensus code we have, I am going to have to be a bit nitpicky.

In order urgency:
1. How the provider is initialized and whether it even needs to be. There are some panics and some other obvious issues.
2. PeginMeta ergonomics considering future versions
3. Lack of deser/ser unit tests
4. Lack of integration tests. This doesnt have to neccecarily impede the merge of this work but certainly the release of it.  
5. Unneccecary changes in `poa/mod.rs`

SatwikPrabhu commented:
> This is on the right track and I commend you for pioneering us into a new pegin version! But I did notice a couple things that stuck out and since this is touching probably the most critical consensus code we have, I am going to have to be a bit nitpicky.
> 
> In order urgency:
> 
> How the provider is initialized and whether it even needs to be. There are some panics and some other obvious issues.
> PeginMeta ergonomics considering future versions
> Lack of deser/ser unit tests
> Lack of integration tests. This doesnt have to neccecarily impede the merge of this work but certainly the release of it.
> Unneccecary changes in poa/mod.rs

@0xBEEFCAF3 Thanks for the extensive review. Appreciate it. I agree that this needs some refactoring and will do my best to  get this done today.

SatwikPrabhu commented:
> > This is on the right track and I commend you for pioneering us into a new pegin version! But I did notice a couple things that stuck out and since this is touching probably the most critical consensus code we have, I am going to have to be a bit nitpicky.
> 
> In order urgency:
> 
> How the provider is initialized and whether it even needs to be. There are some panics and some other obvious issues.
> PeginMeta ergonomics considering future versions
> Lack of deser/ser unit tests
> Lack of integration tests. This doesnt have to neccecarily impede the merge of this work but certainly the release of it.
> Unneccecary changes in poa/mod.rs

I've addressed these comments. There are still some things to do:

1. Need to provide the correct hex for `deserialize_pegin_metadata_v1` to match the new pegin meta v1
2. I've made a stupid mistake here by squashing the main commit of this pr with a commit outside of this PR. Need to fix that somehow because it will be very difficult to review otherwise
3. Fix clippy errors and rebase conflicts
4. Add documentation for some new helper functions that I added
5. Add integration test. As mentioned does not block merging this PR

Feel free to add any other things in the above todo list.
